diff -ru gcc-5.2.0/gcc/cfgbuild.c /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/cfgbuild.c
--- gcc-5.2.0/gcc/cfgbuild.c	2015-01-15 14:28:42.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/cfgbuild.c	2016-02-09 14:56:31.261691823 +0100
@@ -91,6 +91,9 @@
     case NOTE:
       return false;
 
+    case DIABLO_SWITCHINS:
+      return false;
+
     default:
       gcc_unreachable ();
     }
@@ -141,6 +144,9 @@
 	 this may happen.  */
       return false;
 
+    case DIABLO_SWITCHINS:
+      return false;
+
     default:
       gcc_unreachable ();
     }
diff -ru gcc-5.2.0/gcc/cfgrtl.c /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/cfgrtl.c
--- gcc-5.2.0/gcc/cfgrtl.c	2015-04-14 15:25:34.000000000 +0200
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/cfgrtl.c	2016-02-09 14:56:31.261691823 +0100
@@ -4225,6 +4225,11 @@
 	      gcc_unreachable ();
 	    }
 	  break;
+
+        case DIABLO_SWITCHINS:
+          emit_diablo_switchins_copy(insn);
+          break;
+
 	default:
 	  gcc_unreachable ();
 	}
diff -ru gcc-5.2.0/gcc/emit-rtl.c /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/emit-rtl.c
--- gcc-5.2.0/gcc/emit-rtl.c	2015-01-23 16:57:42.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/emit-rtl.c	2016-02-09 14:56:31.261691823 +0100
@@ -2561,6 +2561,16 @@
 				NULL, label_num++, NULL));
 }
 
+
+static uint32_t diablo_switchins_num = 0;
+
+rtx_code_label *
+gen_diablo_switchins (void)
+{
+  return as_a <rtx_code_label *> (gen_rtx_DIABLO_SWITCHINS (VOIDmode, 0, NULL_RTX, NULL_RTX, NULL,
+                                   diablo_switchins_num++, 0));
+}
+
 /* For procedure integration.  */
 
 /* Install new pointers to the first and last insns in the chain.
@@ -5114,6 +5124,27 @@
   return jump_table_data;
 }
 
+rtx_insn *
+emit_diablo_switchins(rtx a)
+{
+  gcc_checking_assert (INSN_UID (a) == 0);
+  INSN_UID(a) = cur_insn_uid++;
+  add_insn(as_a <rtx_insn *> (a));
+  return as_a <rtx_insn *> (a);
+}
+
+rtx_insn *
+emit_diablo_switchins_copy(rtx orig)
+{
+  rtx_code_label *ret = gen_diablo_switchins();
+  DIABLO_SWITCHINS_TABLEID(ret) = DIABLO_SWITCHINS_TABLEID(orig);
+
+  INSN_UID(ret) = cur_insn_uid++;
+  add_insn(as_a <rtx_insn *> (ret));
+
+  return as_a <rtx_insn *> (ret);
+}
+
 /* Make an insn of code BARRIER
    and add it to the end of the doubly-linked list.  */
 
diff -ru gcc-5.2.0/gcc/emit-rtl.h /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/emit-rtl.h
--- gcc-5.2.0/gcc/emit-rtl.h	2015-01-25 15:26:59.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/emit-rtl.h	2016-02-09 14:56:31.261691823 +0100
@@ -189,4 +189,11 @@
    MODE and adjusted by OFFSET.  */
 extern rtx widen_memory_access (rtx, machine_mode, HOST_WIDE_INT);
 
+rtx_insn *
+emit_diablo_switchins(rtx a);
+rtx_code_label *
+gen_diablo_switchins (void);
+rtx_insn *
+emit_diablo_switchins_copy(rtx orig);
+
 #endif /* GCC_EMIT_RTL_H */
diff -ru gcc-5.2.0/gcc/final.c /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/final.c
--- gcc-5.2.0/gcc/final.c	2015-01-15 14:28:42.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/final.c	2016-02-09 15:39:24.135165897 +0100
@@ -1156,7 +1156,8 @@
       INSN_ADDRESSES (uid) = insn_current_address + insn_lengths[uid];
 
       if (NOTE_P (insn) || BARRIER_P (insn)
-	  || LABEL_P (insn) || DEBUG_INSN_P (insn))
+	  || LABEL_P (insn) || DEBUG_INSN_P (insn)
+          || DIABLO_P(insn))
 	continue;
       if (insn->deleted ())
 	continue;
@@ -2417,6 +2418,14 @@
     case BARRIER:
       break;
 
+    case DIABLO_SWITCHINS:
+    {
+        char diablobuf[128];
+        sprintf(diablobuf, "DiabloSwitchIns%d_", DIABLO_SWITCHINS_ID(insn));
+        targetm.asm_out.internal_label (file, diablobuf, DIABLO_SWITCHINS_TABLEID(insn));
+    }
+      break;
+
     case CODE_LABEL:
       /* The target port might emit labels in the output function for
 	 some insn, e.g. sh.c output_branchy_insn.  */
@@ -2475,6 +2484,10 @@
 	  else
 	    switch_to_section (current_function_section ());
 
+          char diablobuf[128];
+          sprintf(diablobuf, "DiabloSwitchTable%d_", CODE_LABEL_NUMBER(insn));
+          targetm.asm_out.internal_label (file, diablobuf, XVECLEN (PATTERN(next), GET_CODE (PATTERN(next)) == ADDR_DIFF_VEC));
+
 #ifdef ASM_OUTPUT_CASE_LABEL
 	  ASM_OUTPUT_CASE_LABEL (file, "L", CODE_LABEL_NUMBER (insn),
 				 next);
diff -ru gcc-5.2.0/gcc/haifa-sched.c /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/haifa-sched.c
--- gcc-5.2.0/gcc/haifa-sched.c	2015-02-19 09:31:14.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/haifa-sched.c	2016-02-09 15:00:31.341399952 +0100
@@ -1929,6 +1929,7 @@
 /* Index POINT gives the instruction at point POINT of the model schedule.
    This array doesn't change during main scheduling.  */
 static vec<rtx_insn *> model_schedule;
+static vec<rtx_insn *> diablo_insns;
 
 /* The list of instructions in the model worklist, sorted in order of
    decreasing priority.  */
@@ -6426,6 +6427,36 @@
       && sched_pressure != SCHED_PRESSURE_MODEL && !sched_fusion)
     find_modifiable_mems (head, tail);
 
+ int nr_diablo_ins = 0;
+/* unlink diablo switch instructions */
+ {
+        rtx_insn *b;
+        for (b = head; b != NEXT_INSN(tail); b = NEXT_INSN(b))
+          if (GET_CODE(b) == DIABLO_SWITCHINS)
+               nr_diablo_ins++;
+
+        if (nr_diablo_ins > 0) {
+                diablo_insns.create(nr_diablo_ins);
+
+                for (b = head; b != NEXT_INSN(tail); b = NEXT_INSN(b)) {
+                        if (DIABLO_P(b)) {
+                        	gcc_assert(GET_CODE(b) == DIABLO_SWITCHINS);
+                                rtx_insn * p = PREV_INSN(b);
+                                rtx_insn * n = NEXT_INSN(b);
+
+                                SET_NEXT_INSN(p) = n;
+                                SET_PREV_INSN(n) = p;
+
+                                if (b == head)
+                                        head = n;
+                                gcc_assert(b != tail);
+
+                                diablo_insns.quick_push(b);
+                        }
+                }
+        }
+}
+
   /* We used to have code to avoid getting parameters moved from hard
      argument registers into pseudos.
 
@@ -7060,6 +7091,28 @@
 
   head = restore_other_notes (head, NULL);
 
+  if (nr_diablo_ins > 0)
+  {
+   rtx_insn * diablo_insn;
+   unsigned int i;
+
+   FOR_EACH_VEC_ELT(diablo_insns, i, diablo_insn) {
+                rtx_insn * n = NEXT_INSN(diablo_insn);
+                rtx_insn * p = PREV_INSN(n);
+
+                SET_NEXT_INSN(diablo_insn) = n;
+                SET_PREV_INSN(diablo_insn) = p;
+
+                SET_NEXT_INSN(p) = diablo_insn;
+                SET_PREV_INSN(n) = diablo_insn;
+
+                if (head == n)
+                        head = diablo_insn;
+        }
+    diablo_insns.release();
+  }
+
+
   current_sched_info->head = head;
   current_sched_info->tail = tail;
 
@@ -7780,7 +7833,7 @@
 static int
 haifa_luid_for_non_insn (rtx x)
 {
-  gcc_assert (NOTE_P (x) || LABEL_P (x));
+  gcc_assert (NOTE_P (x) || LABEL_P (x) || DIABLO_P(x));
 
   return 0;
 }
diff -ru gcc-5.2.0/gcc/jump.c /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/jump.c
--- gcc-5.2.0/gcc/jump.c	2015-01-27 10:19:30.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/jump.c	2016-02-09 14:56:31.265692018 +0100
@@ -361,6 +361,8 @@
 	{
 	  if (insn->deleted ())
 	    ;
+          else if (DIABLO_P(insn))
+                ;
 	  else if (LABEL_P (insn))
 	    prev_nonjump_insn = NULL;
 	  else if (JUMP_TABLE_DATA_P (insn))
diff -ru gcc-5.2.0/gcc/rtl.def /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/rtl.def
--- gcc-5.2.0/gcc/rtl.def	2015-01-05 13:33:28.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/rtl.def	2016-02-09 14:56:31.265692018 +0100
@@ -761,6 +761,8 @@
    been optimized away completely.  */
 DEF_RTL_EXPR(DEBUG_PARAMETER_REF, "debug_parameter_ref", "t", RTX_OBJ)
 
+DEF_RTL_EXPR(DIABLO_SWITCHINS, "diablo_switchins", "iuuBii", RTX_EXTRA)
+
 /* All expressions from this point forward appear only in machine
    descriptions.  */
 #ifdef GENERATOR_FILE
diff -ru gcc-5.2.0/gcc/rtl.h /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/rtl.h
--- gcc-5.2.0/gcc/rtl.h	2015-01-05 23:04:53.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/rtl.h	2016-02-09 14:56:31.265692018 +0100
@@ -758,6 +758,8 @@
 /* Predicate yielding nonzero iff X is a label insn.  */
 #define LABEL_P(X) (GET_CODE (X) == CODE_LABEL)
 
+#define DIABLO_P(X) (GET_CODE(X) == DIABLO_SWITCHINS)
+
 /* Predicate yielding nonzero iff X is a jump insn.  */
 #define JUMP_P(X) (GET_CODE (X) == JUMP_INSN)
 
@@ -1418,6 +1420,9 @@
     return XVEC (pat, 1); /* presumably an ADDR_DIFF_VEC */
 }
 
+#define DIABLO_SWITCHINS_ID(X) XINT(X, 4)
+#define DIABLO_SWITCHINS_TABLEID(X) XINT(X, 5)
+
 #define RTX_FRAME_RELATED_P(RTX)					\
   (RTL_FLAG_CHECK6 ("RTX_FRAME_RELATED_P", (RTX), DEBUG_INSN, INSN,	\
 		    CALL_INSN, JUMP_INSN, BARRIER, SET)->frame_related)
diff -ru gcc-5.2.0/gcc/sched-rgn.c /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/sched-rgn.c
--- gcc-5.2.0/gcc/sched-rgn.c	2015-01-05 13:33:28.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/sched-rgn.c	2016-02-09 14:56:31.265692018 +0100
@@ -2492,10 +2492,10 @@
 #endif
 		 || (!reload_completed
 		     && sets_likely_spilled (PATTERN (insn)))))
-	 || NOTE_P (insn)
+	 || NOTE_P (insn) || DIABLO_P(insn)
 	 || (last != 0 && SCHED_GROUP_P (last)))
     {
-      if (!NOTE_P (insn))
+      if (!NOTE_P (insn) && !DIABLO_P(insn))
 	{
 	  if (last != 0
 	      && sd_find_dep_between (insn, last, false) == NULL)
@@ -2527,7 +2527,7 @@
 	insn = prev_nonnote_insn (insn);
 
 	if (bitmap_bit_p (insn_referenced, INSN_LUID (insn))
-	    || DEBUG_INSN_P (insn))
+	    || DEBUG_INSN_P (insn) || DIABLO_P(insn))
 	  continue;
 
 	if (! sched_insns_conditions_mutex_p (last, insn))
diff -ru gcc-5.2.0/gcc/stmt.c /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/stmt.c
--- gcc-5.2.0/gcc/stmt.c	2015-02-13 15:42:30.000000000 +0100
+++ /home/jens/diablo/repos/diablo-toolchains/crosstool-ng/.build/src/gcc-5.2.0/gcc/stmt.c	2016-02-09 15:06:46.711700038 +0100
@@ -81,6 +81,8 @@
 #include "dumpfile.h"
 #include "builtins.h"
 
+static rtx_code_label * diablo_table_label;
+
 
 /* Functions and data structures for expanding case statements.  */
 
@@ -1098,6 +1100,7 @@
     }
   /* Output the table.  */
   emit_label (table_label);
+  diablo_table_label = table_label;
 
   if (CASE_VECTOR_PC_RELATIVE || flag_pic)
     emit_jump_table_data (gen_rtx_ADDR_DIFF_VEC (CASE_VECTOR_MODE,
@@ -1271,10 +1274,14 @@
     emit_case_decision_tree (index_expr, index_type,
                              case_list, default_label,
                              default_prob);
-  else
+  else {
+    rtx_code_label *diablo_switchins = gen_diablo_switchins();
+    emit_label_after(diablo_switchins, before_case);
     emit_case_dispatch_table (index_expr, index_type,
 			      case_list, default_label,
 			      minval, maxval, range, bb);
+    DIABLO_SWITCHINS_TABLEID(diablo_switchins) = CODE_LABEL_NUMBER(diablo_table_label);
+  }
 
   reorder_insns (NEXT_INSN (before_case), get_last_insn (), before_case);
 
@@ -1363,10 +1370,13 @@
 	  case_list = add_case_node (case_list, low, low, lab, 0, case_node_pool);
 	}
 
+      rtx diablo_switchins = gen_diablo_switchins();
+      emit_label_after(diablo_switchins, before_case);
       emit_case_dispatch_table (index_expr, index_type,
 				case_list, default_label,
 				minval, maxval, range,
                                 BLOCK_FOR_INSN (before_case));
+      DIABLO_SWITCHINS_TABLEID(diablo_switchins) = CODE_LABEL_NUMBER(diablo_table_label);
       emit_label (default_label);
       free_alloc_pool (case_node_pool);
     }
